File: app-android/src/main/java/com/ustadmobile/port/android/view/Register2Activity.java
package com.ustadmobile.port.android.view;

import android.os.Bundle;
import android.support.design.widget.TextInputEditText;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.Register2Presenter;
import com.ustadmobile.core.generated.locale.MessageID;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.core.view.Register2View;
import com.ustadmobile.lib.db.entities.Person;
import com.ustadmobile.port.android.util.UMAndroidUtil;

import java.util.HashMap;


public class Register2Activity extends UstadBaseActivity implements Register2View {


    private TextView errorMessage;

    private UstadMobileSystemImpl systemImpl;

    private HashMap<Integer, Integer> fieldToViewIdMap = new HashMap<>();

    private Register2Presenter presenter;

    private String serverUrl;

    private ProgressBar progressDialog;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register2);

        fieldToViewIdMap.put(FIELD_FIRST_NAME, R.id.activity_create_account_firstname_text);
        fieldToViewIdMap.put(FIELD_LAST_NAME, R.id.activity_create_account_lastname_text);
        fieldToViewIdMap.put(FIELD_EMAIL, R.id.activity_create_account_email_text);
        fieldToViewIdMap.put(FIELD_USERNAME, R.id.activity_create_account_username_text);
        fieldToViewIdMap.put(FIELD_PASSWORD, R.id.activity_create_account_password_text);
        fieldToViewIdMap.put(FIELD_CONFIRM_PASSWORD, R.id.activity_create_account_password_confirmpassword_text);

        Button registerUser = findViewById(R.id.activity_create_account_create_account_button);
        errorMessage = findViewById(R.id.activity_create_account_error_text);
        progressDialog = findViewById(R.id.progressBar);
        progressDialog.setIndeterminate(true);

        systemImpl = UstadMobileSystemImpl.getInstance();

        presenter = new Register2Presenter(this,
                UMAndroidUtil.bundleToHashtable(getIntent().getExtras()),this);
        presenter.onCreate(UMAndroidUtil.bundleToHashtable(savedInstanceState));

        registerUser.setOnClickListener(v -> checkAccountFields());

    }


    private void checkAccountFields(){
        for(Integer fieldCode: fieldToViewIdMap.values()){
            if(getFieldValue(fieldCode).isEmpty()){
                setErrorMessage(systemImpl.getString(MessageID.register_empty_fields,this));
                return;
            }
        }

        if(getFieldValue(FIELD_PASSWORD).equals(getFieldValue(FIELD_CONFIRM_PASSWORD))){

            String emailPattern = "[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+";

            if(getFieldValue(FIELD_EMAIL).matches(emailPattern)){
                if(getFieldValue(FIELD_PASSWORD).length() <= 5){
                    setErrorMessage(systemImpl.getString(MessageID.field_password_min,this));
                }else{
                    Person person = new Person();
                    person.setFirstNames(getFieldValue(FIELD_FIRST_NAME));
                    person.setLastName(getFieldValue(FIELD_LAST_NAME));
                    person.setEmailAddr(getFieldValue(FIELD_EMAIL));
                    person.setUsername(getFieldValue(FIELD_USERNAME));
                    presenter.handleClickRegister(person,getFieldValue(FIELD_PASSWORD),serverUrl);
                }
            }else{
                setErrorMessage(systemImpl.getString(MessageID.register_incorrect_email,this));
            }
        }else{
            setErrorMessage(systemImpl.getString(MessageID.filed_password_no_match,this));
        }
    }



    @Override
    public void setErrorMessage(String errorMessage) {
        this.errorMessage.setText(errorMessage);

    }

    @Override
    public void setServerUrl(String url) {
        this.serverUrl = url;
    }

    @Override
    public void setInProgress(boolean inProgress) {
        progressDialog.setVisibility(inProgress ? View.VISIBLE:View.GONE);
    }

    private String getFieldValue(int fieldCode) {
        return ((TextInputEditText)findViewById(fieldToViewIdMap.get(fieldCode))).getText().toString();
    }


}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/Register2Activity.kt
package com.ustadmobile.port.android.view

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.google.android.material.textfield.TextInputEditText
import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.Register2Presenter
import com.ustadmobile.core.generated.locale.MessageID
import com.ustadmobile.core.impl.UMAndroidUtil.bundleToMap
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.core.view.Register2View
import com.ustadmobile.lib.db.entities.Person
import java.util.*


class Register2Activity : UstadBaseActivity(), Register2View {


    private var errorMessageView: TextView? = null

    private var systemImpl: UstadMobileSystemImpl? = null

    private val fieldToViewIdMap = HashMap<Int, Int>()

    private var presenter: Register2Presenter? = null

    private var serverUrl: String? = null

    private var registerUser: Button? = null

    private var progressDialog: ProgressBar? = null

    private val textWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            errorMessageView!!.visibility = View.GONE
            checkRegisterButtonStatus()
        }

        override fun afterTextChanged(s: Editable) {}
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register2)

        setUMToolbar(R.id.um_toolbar)
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setHomeButtonEnabled(true)
        }
        fieldToViewIdMap[Register2View.FIELD_FIRST_NAME] = R.id.activity_create_account_firstname_text
        fieldToViewIdMap[Register2View.FIELD_LAST_NAME] = R.id.activity_create_account_lastname_text
        fieldToViewIdMap[Register2View.FIELD_EMAIL] = R.id.activity_create_account_email_text
        fieldToViewIdMap[Register2View.FIELD_USERNAME] = R.id.activity_create_account_username_text
        fieldToViewIdMap[Register2View.FIELD_PASSWORD] = R.id.activity_create_account_password_text
        fieldToViewIdMap[Register2View.FIELD_CONFIRM_PASSWORD] = R.id.activity_create_account_password_confirmpassword_text

        registerUser = findViewById(R.id.activity_create_account_create_account_button)
        errorMessageView = findViewById(R.id.activity_create_account_error_text)
        progressDialog = findViewById(R.id.progressBar)
        progressDialog!!.isIndeterminate = true
        progressDialog!!.scaleY = 3f

        systemImpl = UstadMobileSystemImpl.instance

        presenter = Register2Presenter(this,
                bundleToMap(intent.extras), this,
                UmAccountManager.getRepositoryForActiveAccount(this).personDao)
        presenter!!.onCreate(bundleToMap(savedInstanceState))

        for (fieldId in fieldToViewIdMap.values) {
            (findViewById<View>(fieldId) as TextInputEditText).addTextChangedListener(textWatcher)
        }

        registerUser!!.setOnClickListener { v -> checkAccountFields() }

    }


    private fun checkAccountFields() {
        for (fieldCode in fieldToViewIdMap.keys) {
            if (getFieldValue(fieldCode).isEmpty()) {
                setErrorMessageView(systemImpl!!.getString(MessageID.register_empty_fields, this))
                return
            }
        }

        checkRegisterButtonStatus()

        if (getFieldValue(Register2View.FIELD_PASSWORD) == getFieldValue(Register2View.FIELD_CONFIRM_PASSWORD)) {

            val emailPattern = "[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+"

            if (getFieldValue(Register2View.FIELD_EMAIL).matches(emailPattern.toRegex())) {
                if (getFieldValue(Register2View.FIELD_PASSWORD).length < 5) {
                    disableButton(true)
                    setErrorMessageView(systemImpl!!.getString(MessageID.field_password_error_min, this))
                } else {
                    val person = Person()
                    person.firstNames = getFieldValue(Register2View.FIELD_FIRST_NAME)
                    person.lastName = getFieldValue(Register2View.FIELD_LAST_NAME)
                    person.emailAddr = getFieldValue(Register2View.FIELD_EMAIL)
                    person.username = getFieldValue(Register2View.FIELD_USERNAME)
                    Thread {
                        presenter!!.handleClickRegister(person,
                                getFieldValue(Register2View.FIELD_PASSWORD), serverUrl!!)
                    }.start()
                }
            } else {
                disableButton(true)
                setErrorMessageView(systemImpl!!.getString(MessageID.register_incorrect_email, this))
            }
        } else {
            disableButton(true)
            setErrorMessageView(systemImpl!!.getString(MessageID.filed_password_no_match, this))
        }
    }

    private fun disableButton(disable: Boolean) {
        registerUser!!.setBackgroundColor(ContextCompat.getColor(this,
                if (disable) R.color.divider else R.color.accent))
        registerUser!!.isEnabled = !disable
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        if (item.itemId == android.R.id.home) {
            finish()
        }
        return true
    }

    override fun setErrorMessageView(errorMessageView: String) {
        this.errorMessageView!!.text = errorMessageView
        this.errorMessageView!!.visibility = View.VISIBLE
        disableButton(true)
    }

    override fun setServerUrl(url: String) {
        this.serverUrl = url
    }

    override fun setInProgress(inProgress: Boolean) {
        progressDialog!!.visibility = if (inProgress) View.VISIBLE else View.GONE
    }

    private fun getFieldValue(fieldCode: Int): String {
        return (findViewById<View>(fieldToViewIdMap[fieldCode]!!) as TextInputEditText).text!!.toString()
    }

    private fun checkRegisterButtonStatus() {
        var isEnabled = true
        for (fieldCode in fieldToViewIdMap.keys) {
            if (getFieldValue(fieldCode).isEmpty()) {
                isEnabled = false
                break
            }
        }
        disableButton(!isEnabled)
    }


}
